OMCompiler/Compiler/NBackEnd/NBackendDAE.mo

```diff
@@ -0,0 +1,165 @@
+/*
+* This file is part of OpenModelica.
+*
+* Copyright (c) 1998-2014, Open Source Modelica Consortium (OSMC),
+* c/o Linköpings universitet, Department of Computer and Information Science,
+* SE-58183 Linköping, Sweden.
+*
+* All rights reserved.
+*
+* THIS PROGRAM IS PROVIDED UNDER THE TERMS OF GPL VERSION 3 LICENSE OR
+* THIS OSMC PUBLIC LICENSE (OSMC-PL) VERSION 1.2.
+* ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS PROGRAM CONSTITUTES
+* RECIPIENT'S ACCEPTANCE OF THE OSMC PUBLIC LICENSE OR THE GPL VERSION 3,
+* ACCORDING TO RECIPIENTS CHOICE.
+*
+* The OpenModelica software and the Open Source Modelica
+* Consortium (OSMC) Public License (OSMC-PL) are obtained
+* from OSMC, either from the above address,
+* from the URLs: http://www.ida.liu.se/projects/OpenModelica or
+* http://www.openmodelica.org, and in the OpenModelica distribution.
+* GNU version 3 is obtained from: http://www.gnu.org/copyleft/gpl.html.
+*
+* This program is distributed WITHOUT ANY WARRANTY; without
+* even the implied warranty of  MERCHANTABILITY or FITNESS
+* FOR A PARTICULAR PURPOSE, EXCEPT AS EXPRESSLY SET FORTH
+* IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE CONDITIONS OF OSMC-PL.
+*
+* See the full OSMC Public License conditions for more details.
+*
+*/
+encapsulated package NBackendDAE
+" file:        NBackendDAE.mo
+ package:     NBackendDAE
+ description: This file contains the data-types used by the back end.
+"
+ uniontype BackendStructure
+ record BDAE
+   /* Stuff here! */
+ end BDAE;
+end BackendStructure;
+
+/* ========================================================
+               ADJACENCY MATRIX AND MATCHING
+   ======================================================== */
+
+/*
+  Regular slice. Always has three elements.
+  E.g. Start=1, Stop=91, Step=3
+  =>  {1,4,7,...,88,91}
+  The order matters for eq <=> var matchings!
+  [1,91,3] <> [91,1,-3]
+*/
+type RegularSlice = array<Integer>;
+
+/*
+  Vector slice type (one dimension). Contains a list
+  of static singleton indices and a list of regular
+  slices. Each regular slice needs to represent the
+  same number of (scalarized) element for it to be
+  consistent. The singletons are assumed to be static
+  and occur for every scalarized instance of the
+  regular slices. E.g.
+  for i in 1:3 loop
+    x[i] = x[4];
+  end for;
+  => ({4}, {[1,3,1]})
+*/
+uniontype VectorSlice
```

Comment:
Why do we mix list and array here? 

Are we going to mutate this structure alot? Will there be random access? 

DoubleEndedList might be better

Reply:
The order does not matter and an element never has to be accessed without traversing all of them (as far as i know right now). Also the number of elements may change a lot. I don't see the advantage of using a double ended list right now, but i also don't see a disadvantage. So maybe i should change it.

Reply:
And the inner part being an array comes from the fact that it always will have three elements with specific meanings. Since it will be accessed quite often i thought a straight up array is more effective than having a record with three elements.

Reply:
It does sound a bit like the mutable double-ended list might be a good option (Just looking through things in general)